bittensor/_dataset/dataset_impl.py

```diff
@@ -1,536 +1,490 @@
-            elif action == 'post':
-                response = self.requests_retry_session(session=session).post(address, timeout=timeout)
-            logger.success("Loaded from IPFS:".ljust(20) + "<blue>{}</blue>".format(file_meta['Name']))
+        self.max_threads = max_threads
+        self._idle_semaphore = threading.Semaphore(0)
+        self._threads = []
+        self._shutdown_lock = threading.Lock()
+        self._shutdown = False
+
+    def submit(self, fn, args:list=[],kwargs:dict={}):
+        '''
+        Submit a function with args and kwargs on a seperate thread.
+
+        Args
+            fn (Callable):
+                Function to place on the thread.
+            args (list):
+                Arguments to a function.
+            kwargs (dict):
+                Key word arguments to a function.
+        '''
+        with self._shutdown_lock:
+            if self._shutdown:
+                raise RuntimeError('cannot schedule new futures after shutdown')
+            
+            thread = threading.Thread(target=fn, args=args, kwargs=kwargs, daemon=True)
+            thread.start()
+            self._threads.append(thread)
 
-        except Exception as E:
-            logger.error(f"Failed to get from IPFS {file_meta['Name']} {E}")
-            return None
+        return thread
 
-        return response
 
-    def __len__(self):
-        """ Returns length of the dataset that the dataset is processing
-        """
+    @property
+    def threads(self):
+        '''List threads.'''
+        return self._threads
+
+    def __del__(self):
+        self.shutdown()
+
+    def shutdown(self, wait=True):
+        '''Shutdown threads'''
+        if wait:
+            for t in self._threads:
+                t.join()
+
 
-    def __getitem__(self, idx):
-        """ Returns the next batch from the dataset.
-        """
 
-class GenesisTextDataset( Dataset ):
-    """ One kind of dataset that caters for the data from ipfs 
+
+class GenesisTextDataset:
+    """ Implementation for the dataset class, which handles dataloading from ipfs
     """
+
+    ipfs_url = 'http://global.ipfs.opentensor.ai/api/v0'
+    dataset_dir = 'http://global.ipfs.opentensor.ai/api/v0/cat' 
+    text_dir = 'http://global.ipfs.opentensor.ai/api/v0/object/get'
+    mountain_hash = 'QmSdDg6V9dgpdAFtActs75Qfc36qJtm9y8a7yrQ1rHm7ZX'
+
     def __init__(
-        self,
-        block_size,
-        batch_size,
-        num_workers,
-        dataset_name,
-        data_dir,
-        save_dataset,
-        max_datasets,
-        no_tokenizer, 
-        num_batches,
-        max_directories
-    ):
-        super().__init__()
-        self.block_size = block_size
+            self, 
+            batch_size: int=8, 
+            sequence_length: int=256,
+            block_size: int = 10000,
+            num_workers: int = 1,
+            dataset_name: List[str]=['ArXiv'], 
```

Comment:
Same as the other comment, default args shouldn't be mutable.

Reply:
Also, consider using `typing.Sequence` if it doesn't *have* to be a list (i.e. any reason i can't throw a tuple?)